File: internal/channels.go
package internal

import (
	"net/http"
	"sort"
	"strings"
	"sync"
)

const (
	DefaultChannelRetentionMs             int64 = 24 * 60 * 60 * 1000
	DefaultChannelInactivityTtlMs         int64 = 0
	DefaultChannelKeepaliveIntervalMs     int64 = 25 * 1000
	DefaultChannelMaxConnectionLifetimeMs int64 = 2 * 60 * 60 * 1000
)

type ChannelTransport string

const (
	ChannelTransportWS ChannelTransport = "ws"
)

type ChannelOperation string

const (
	ChannelOperationSubscribe ChannelOperation = "subscribe"
	ChannelOperationPublish   ChannelOperation = "publish"
	ChannelOperationConnect   ChannelOperation = "connect"
)

type ChannelLifecycleConfig struct {
	ReplayWindowMs          int64 `json:"replayWindowMs,omitempty"`
	InactivityTtlMs         int64 `json:"inactivityTtlMs"`
	KeepaliveIntervalMs     int64 `json:"keepaliveIntervalMs,omitempty"`
	MaxConnectionLifetimeMs int64 `json:"maxConnectionLifetimeMs,omitempty"`
}

func (c ChannelLifecycleConfig) withDefaults() ChannelLifecycleConfig {
	if c.ReplayWindowMs == 0 {
		c.ReplayWindowMs = DefaultChannelRetentionMs
	}
	if c.KeepaliveIntervalMs == 0 {
		c.KeepaliveIntervalMs = DefaultChannelKeepaliveIntervalMs
	}
	if c.MaxConnectionLifetimeMs == 0 {
		c.MaxConnectionLifetimeMs = DefaultChannelMaxConnectionLifetimeMs
	}
	return c
}

type ChannelAuthContext struct {
	Channel   string           `json:"channel"`
	Operation ChannelOperation `json:"operation"`
	Pattern   string           `json:"pattern"`
}

type ChannelGrant struct {
	Subject string `json:"subject,omitempty"`
	ChannelLifecycleConfig
}

type ChannelAuthDecision struct {
	OK bool `json:"ok"`
	ChannelGrant
}

func AllowChannel(grant ChannelGrant) ChannelAuthDecision {
	grant.ChannelLifecycleConfig = grant.ChannelLifecycleConfig.withDefaults()
	return ChannelAuthDecision{
		OK:           true,
		ChannelGrant: grant,
	}
}

func RejectChannel() ChannelAuthDecision {
	return ChannelAuthDecision{OK: false}
}

type ChannelDefinition struct {
	Auth func(*http.Request, ChannelAuthContext) ChannelAuthDecision
	ChannelLifecycleConfig
	Transport ChannelTransport
}

type ChannelAuthRequest struct {
	URL     string            `json:"url"`
	Method  string            `json:"method,omitempty"`
	Headers map[string]string `json:"headers,omitempty"`
}

type ChannelAuthorizeInput struct {
	Channel   string             `json:"channel"`
	Operation ChannelOperation   `json:"operation"`
	Request   ChannelAuthRequest `json:"request"`
}

type ChannelAuthorizeResponse struct {
	OK        bool             `json:"ok"`
	Transport ChannelTransport `json:"transport,omitempty"`
	ChannelGrant
}

type Channel struct {
	name      string
	transport ChannelTransport
}

func (c *Channel) Name() string {
	return c.name
}

func (c *Channel) Transport() ChannelTransport {
	return c.transport
}

type channelDefinitionEntry struct {
	definition ChannelDefinition
	index      int
	pattern    string
}

type ChannelRegistry struct {
	mu          sync.RWMutex
	definitions []channelDefinitionEntry
	nextIndex   int
}

func NewChannelRegistry() *ChannelRegistry {
	return &ChannelRegistry{}
}

var Channels = NewChannelRegistry()

func (r *ChannelRegistry) Create(name string, definition ...ChannelDefinition) *Channel {
	if len(definition) > 0 {
		r.Define(name, definition[0])
		return &Channel{name: name, transport: definition[0].Transport}
	}
	return &Channel{name: name}
}

func (r *ChannelRegistry) Define(pattern string, definition ChannelDefinition) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.definitions = append(r.definitions, channelDefinitionEntry{
		definition: definition,
		index:      r.nextIndex,
		pattern:    pattern,
	})
	r.nextIndex++
}

func (r *ChannelRegistry) Clear() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.definitions = nil
	r.nextIndex = 0
}

func (r *ChannelRegistry) ResolveDefinition(channel string) *ChannelDefinition {
	r.mu.RLock()
	defer r.mu.RUnlock()

	matches := make([]channelDefinitionEntry, 0)
	for _, entry := range r.definitions {
		if patternMatches(entry.pattern, channel) {
			matches = append(matches, entry)
		}
	}
	if len(matches) == 0 {
		return nil
	}

	sort.SliceStable(matches, func(i, j int) bool {
		left, right := matches[i], matches[j]
		if isExactPattern(left.pattern) != isExactPattern(right.pattern) {
			return isExactPattern(left.pattern)
		}
		leftSpecificity := patternSpecificity(left.pattern)
		rightSpecificity := patternSpecificity(right.pattern)
		if leftSpecificity != rightSpecificity {
			return leftSpecificity > rightSpecificity
		}
		return left.index < right.index
	})

	definition := matches[0].definition
	return &definition
}

func (r *ChannelRegistry) Authorize(input ChannelAuthorizeInput) (ChannelAuthorizeResponse, bool, bool) {
	entry := r.resolveEntry(input.Channel)
	if entry == nil {
		return ChannelAuthorizeResponse{}, false, false
	}

	method := input.Request.Method
	if method == "" {
		method = http.MethodGet
	}
	req, err := http.NewRequest(method, input.Request.URL, nil)
	if err != nil {
		return ChannelAuthorizeResponse{}, true, false
	}
	for key, value := range input.Request.Headers {
		req.Header.Set(key, value)
	}

	decision := entry.definition.Auth(req, ChannelAuthContext{
		Channel:   input.Channel,
		Operation: input.Operation,
		Pattern:   entry.pattern,
	})
	if !decision.OK {
		return ChannelAuthorizeResponse{}, true, false
	}

	grant := decision.ChannelGrant
	if grant.ChannelLifecycleConfig == (ChannelLifecycleConfig{}) {
		grant.ChannelLifecycleConfig = entry.definition.ChannelLifecycleConfig
	}
	grant.ChannelLifecycleConfig = grant.ChannelLifecycleConfig.withDefaults()

	return ChannelAuthorizeResponse{
		OK:           true,
		Transport:    entry.definition.Transport,
		ChannelGrant: grant,
	}, true, true
}

func (r *ChannelRegistry) resolveEntry(channel string) *channelDefinitionEntry {
	r.mu.RLock()
	defer r.mu.RUnlock()

	matches := make([]channelDefinitionEntry, 0)
	for _, entry := range r.definitions {
		if patternMatches(entry.pattern, channel) {
			matches = append(matches, entry)
		}
	}
	if len(matches) == 0 {
		return nil
	}

	sort.SliceStable(matches, func(i, j int) bool {
		left, right := matches[i], matches[j]
		if isExactPattern(left.pattern) != isExactPattern(right.pattern) {
			return isExactPattern(left.pattern)
		}
		leftSpecificity := patternSpecificity(left.pattern)
		rightSpecificity := patternSpecificity(right.pattern)
		if leftSpecificity != rightSpecificity {
			return leftSpecificity > rightSpecificity
		}
		return left.index < right.index
	})

	entry := matches[0]
	return &entry
}

func isExactPattern(pattern string) bool {
	return !strings.Contains(pattern, "*")
}

func patternMatches(pattern, channel string) bool {
	if pattern == "*" {
		return true
	}
	if isExactPattern(pattern) {
		return pattern == channel
	}
	if strings.HasSuffix(pattern, "*") {
		return strings.HasPrefix(channel, strings.TrimSuffix(pattern, "*"))
	}
	return false
}

func patternSpecificity(pattern string) int {
	return len(strings.ReplaceAll(pattern, "*", ""))
}

File: sdk/javascript/src/channels.ts
import type {
  ChannelAuthorizeInput,
  ChannelAuthorizeResponse,
  ChannelConnectOptions,
  ChannelConnection,
  ChannelDefinition,
  ChannelDefinitionTransport,
  ChannelMessage,
  ChannelPublishInput,
  ChannelPublishOptions,
  ChannelSubscribeOptions,
  ChannelSubscription,
} from "./types";

export const TAKO_CHANNELS_BASE_PATH = "/channels";
const DEFAULT_CHANNEL_REPLAY_WINDOW_MS = 24 * 60 * 60 * 1000;
const DEFAULT_CHANNEL_INACTIVITY_TTL_MS = 0;
const DEFAULT_CHANNEL_KEEPALIVE_INTERVAL_MS = 25 * 1000;
const DEFAULT_CHANNEL_MAX_CONNECTION_LIFETIME_MS = 2 * 60 * 60 * 1000;

interface ChannelDefinitionEntry {
  definition: ChannelDefinition;
  index: number;
  pattern: string;
}

function isExactPattern(pattern: string): boolean {
  return !pattern.includes("*");
}

function patternMatches(pattern: string, channel: string): boolean {
  if (pattern === "*") {
    return true;
  }
  if (isExactPattern(pattern)) {
    return pattern === channel;
  }
  if (pattern.endsWith("*")) {
    return channel.startsWith(pattern.slice(0, -1));
  }
  return false;
}

function patternSpecificity(pattern: string): number {
  return pattern.replace(/\*/g, "").length;
}

function normalizeBaseUrl(baseUrl?: string): URL {
  if (baseUrl) {
    return new URL(baseUrl);
  }
  if (typeof globalThis.location?.origin === "string" && globalThis.location.origin.length > 0) {
    return new URL(globalThis.location.origin);
  }
  throw new Error("Channel operations require a baseUrl outside the browser.");
}

function channelBaseUrl(channel: string, baseUrl?: string): URL {
  const url = normalizeBaseUrl(baseUrl);
  url.pathname = `${TAKO_CHANNELS_BASE_PATH}/${encodeURIComponent(channel)}`;
  url.search = "";
  return url;
}

function withQuery(url: URL, key: string, value?: string | number): URL {
  if (value !== undefined) {
    url.searchParams.set(key, String(value));
  }
  return url;
}

function toWebSocketUrl(url: URL): string {
  const ws = new URL(url.toString());
  ws.protocol = ws.protocol === "https:" ? "wss:" : "ws:";
  return ws.toString();
}

function requestHeaders(headers?: Record<string, string>): HeadersInit | undefined {
  if (!headers) {
    return undefined;
  }
  return headers;
}

function buildFetchInit(
  base: Omit<RequestInit, "headers" | "signal">,
  options: { headers?: HeadersInit; signal?: AbortSignal },
): RequestInit {
  const init: RequestInit = { ...base };
  if (options.headers !== undefined) {
    init.headers = options.headers;
  }
  if (options.signal !== undefined) {
    init.signal = options.signal;
  }
  return init;
}

function fetchInitOptions(headers?: HeadersInit, signal?: AbortSignal) {
  const options: { headers?: HeadersInit; signal?: AbortSignal } = {};
  if (headers !== undefined) {
    options.headers = headers;
  }
  if (signal !== undefined) {
    options.signal = signal;
  }
  return options;
}

function defaultEventSourceFactory(url: string): unknown {
  const ctor = globalThis.EventSource;
  if (!ctor) {
    throw new Error("EventSource is not available in this runtime.");
  }
  return new ctor(url);
}

function defaultWebSocketFactory(url: string): unknown {
  const ctor = globalThis.WebSocket;
  if (!ctor) {
    throw new Error("WebSocket is not available in this runtime.");
  }
  return new ctor(url);
}

function closeRaw(raw: unknown, code?: number, reason?: string): void {
  if (typeof raw !== "object" || raw === null) {
    return;
  }
  const maybeClose = (raw as { close?: (code?: number, reason?: string) => void }).close;
  if (typeof maybeClose === "function") {
    maybeClose.call(raw, code, reason);
  }
}

function sendRaw(raw: unknown, data: unknown): void {
  if (typeof raw !== "object" || raw === null) {
    throw new Error("Channel connection does not support send().");
  }
  const maybeSend = (raw as { send?: (data: unknown) => void }).send;
  if (typeof maybeSend !== "function") {
    throw new Error("Channel connection does not support send().");
  }
  let payload = data;
  if (
    data !== null &&
    typeof data === "object" &&
    !(data instanceof ArrayBuffer) &&
    !ArrayBuffer.isView(data) &&
    !(typeof Blob !== "undefined" && data instanceof Blob)
  ) {
    payload = JSON.stringify(data);
  }
  maybeSend.call(raw, payload);
}

export class Channel {
  readonly name: string;
  readonly transport: ChannelDefinitionTransport | undefined;

  constructor(name: string, transport?: ChannelDefinitionTransport) {
    this.name = name;
    this.transport = transport;
  }

  async publish<T = unknown>(
    message: ChannelPublishInput<T>,
    options: ChannelPublishOptions = {},
  ): Promise<ChannelMessage<T>> {
    const url = channelBaseUrl(this.name, options.baseUrl);
    url.pathname = `${url.pathname}/messages`;
    const response = await fetch(url.toString(), {
      ...buildFetchInit(
        {
          method: "POST",
          body: JSON.stringify(message),
        },
        {
          ...fetchInitOptions(
            {
              "Content-Type": "application/json",
              ...options.headers,
            },
            options.signal,
          ),
        },
      ),
    });

    if (!response.ok) {
      throw new Error(`Channel publish failed with status ${response.status}.`);
    }

    return (await response.json()) as ChannelMessage<T>;
  }

  subscribe(options: ChannelSubscribeOptions = {}): ChannelSubscription {
    const url = channelBaseUrl(this.name, options.baseUrl);
    const factory = options.eventSourceFactory ?? defaultEventSourceFactory;
    const init: { headers?: Record<string, string>; lastEventId?: string } = {};
    if (options.headers !== undefined) {
      init.headers = options.headers;
    }
    if (options.lastEventId !== undefined) {
      init.lastEventId = options.lastEventId;
    }
    const raw = factory(url.toString(), init);
    return {
      transport: "sse",
      raw,
      close() {
        closeRaw(raw);
      },
    };
  }

  connect(options: ChannelConnectOptions = {}): ChannelConnection {
    if (this.transport !== "ws") {
      throw new Error("Channel does not enable WebSocket transport.");
    }

    const url = channelBaseUrl(this.name, options.baseUrl);
    withQuery(url, "last_message_id", options.lastMessageId);

    const factory = options.webSocketFactory ?? defaultWebSocketFactory;
    const raw = factory(toWebSocketUrl(url));
    return {
      transport: "ws",
      raw,
      close(code?: number, reason?: string) {
        closeRaw(raw, code, reason);
      },
      send(data: unknown) {
        sendRaw(raw, data);
      },
    };
  }
}

export class ChannelRegistry {
  private definitions: ChannelDefinitionEntry[] = [];
  private nextIndex = 0;

  create(name: string, definition?: ChannelDefinition): Channel {
    if (definition) {
      this.define(name, definition);
      return new Channel(name, definition.transport);
    }
    return new Channel(name);
  }

  define(pattern: string, definition: ChannelDefinition): void {
    this.definitions.push({
      definition,
      index: this.nextIndex++,
      pattern,
    });
  }

  clear(): void {
    this.definitions = [];
    this.nextIndex = 0;
  }

  async authorize(input: ChannelAuthorizeInput): Promise<ChannelAuthorizeResponse> {
    const matched = this.resolveDefinition(input.channel);
    if (!matched) {
      return { ok: false };
    }

    const request = new Request(
      input.request.url,
      buildFetchInit(
        {
          method: input.request.method ?? "GET",
        },
        {
          ...fetchInitOptions(requestHeaders(flattenHeaders(input.request.headers))),
        },
      ),
    );
    const verdict = await matched.definition.auth(request, {
      channel: input.channel,
      operation: input.operation,
      pattern: matched.pattern,
    });

    const config = definitionLifecycleConfig(matched.definition);
    if (verdict === false) {
      return { ok: false };
    }
    if (verdict === true) {
      return { ok: true, ...config };
    }
    return verdict.subject === undefined
      ? { ok: true, ...config }
      : { ok: true, ...config, subject: verdict.subject };
  }

  resolveDefinition(channel: string): ChannelDefinitionEntry | null {
    return (
      this.definitions
        .filter((entry) => patternMatches(entry.pattern, channel))
        .sort((left, right) => {
          const exactWeight =
            Number(isExactPattern(right.pattern)) - Number(isExactPattern(left.pattern));
          if (exactWeight !== 0) {
            return exactWeight;
          }
          const specificity = patternSpecificity(right.pattern) - patternSpecificity(left.pattern);
          if (specificity !== 0) {
            return specificity;
          }
          return left.index - right.index;
        })[0] ?? null
    );
  }
}

function definitionLifecycleConfig(definition: ChannelDefinition) {
  const config: Omit<ChannelAuthorizeResponse, "ok" | "subject"> = {
    replayWindowMs: definition.replayWindowMs ?? DEFAULT_CHANNEL_REPLAY_WINDOW_MS,
    inactivityTtlMs: definition.inactivityTtlMs ?? DEFAULT_CHANNEL_INACTIVITY_TTL_MS,
    keepaliveIntervalMs: definition.keepaliveIntervalMs ?? DEFAULT_CHANNEL_KEEPALIVE_INTERVAL_MS,
    maxConnectionLifetimeMs:
      definition.maxConnectionLifetimeMs ?? DEFAULT_CHANNEL_MAX_CONNECTION_LIFETIME_MS,
  } as ChannelAuthorizeResponse;
  if (definition.transport === "ws") {
    config.transport = "ws";
  }
  return config;
}

function flattenHeaders(
  headers?: Record<string, string | string[]>,
): Record<string, string> | undefined {
  if (!headers) {
    return undefined;
  }
  const flattened: Record<string, string> = {};
  for (const [key, value] of Object.entries(headers)) {
    flattened[key] = Array.isArray(value) ? value.join(", ") : value;
  }
  return flattened;
}
